src/validation/coinbase.cairo

```diff
@@ -66,9 +67,47 @@ pub fn validate_coinbase(
         if witness != @"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0" {
             return Result::Err("Wrong coinbase witness");
         }
-    }
 
-    // TODO: validate BIP-141 segwit output
+        // validate BIP-141 segwit output
+        if *tx.is_segwit {
+            let outputs = *tx.outputs;
+            let mut is_wtxid_commitment_present: bool = false;
+            let mut extracted_wtxid_commitment: ByteArray = "";
+            let mut i = 0;
+            let mut x = 6;
+
+            while i < outputs.len() {
+                let pk_script = *outputs[i].pk_script;
+
+                // check for OP_RETURN and the fixed prefix "0xaa21a9ed"
+                if pk_script.len() >= MINIMUM_WITNESS_COMMITMENT
+                    && pk_script[0] == 0x6a
+                    && pk_script[1] == 0x24
+                    && pk_script[2] == 0xaa
+                    && pk_script[3] == 0x21
+                    && pk_script[4] == 0xa9
+                    && pk_script[5] == 0xed {
+                    // get wtxid commitment
+                    while x < pk_script.len() {
```

Comment:
pk_script can contain trailing non-consensus information, you need to check only 32 bytes (commitment length)